client/src/go/pool/pool.go

```diff
@@ -0,0 +1,241 @@
+	IdleCheckFrequency time.Duration
+	OnConnClosed       func(*Conn) error
+}
+
+func setDefaultOpt(opt Options) Options {
+	return opt
+}
+
+func NewConnPool(opt Options) (*ConnPool, error) {
+	if opt.Dialer == nil {
+		return nil, fmt.Errorf("missing dialing function")
+	}
+	opt = setDefaultOpt(opt)
+	p := &ConnPool{
+		opt:        opt,
+		conns:      make(map[string]*Conn),
+		idleConns:  make([]*Conn, 0, opt.PoolSize),
+		closedChan: make(chan struct{}),
+		connsMu:    new(sync.Mutex),
+		sema:       make(chan struct{}, opt.PoolSize),
+	}
+	p.connsMu.Lock()
+	p.ensureMinIdleConns()
+	p.connsMu.Unlock()
+	if opt.IdleCheckFrequency > 0 && opt.IdleTimeout > 0 {
+		go p.connReaper()
+	}
+
+	return p, nil
+}
+func (p *ConnPool) popIdle() *Conn {
+	if len(p.idleConns) == 0 {
+		return nil
+	}
+	idx := len(p.idleConns) - 1
+	p.totalIdleConns--
+	cn := p.idleConns[idx]
+	p.idleConns = p.idleConns[:idx]
+	p.ensureMinIdleConns()
+	return cn
+}
+
+//TODO
+func (p *ConnPool) Close() {}
+
+func (p *ConnPool) Get() (*Conn, error) {
+	err := p.sema.acquireWithTimeout(p.opt.PoolTimeout)
+	if err != nil {
+		//someplaces are holding connections for so long
+		return nil, err
+	}
+	for {
+		p.connsMu.Lock()
+		cn := p.popIdle() // stack based pool
+		p.connsMu.Unlock()
+		if cn == nil {
+			break //no more idle, create new conn
+		}
+		if p.isStale(cn) {
+			p.connsMu.Lock()
+			//idle connection is always managed, remove here as well
+			delete(p.conns, cn.id)
+			p.totalManagedConns--
+			p.connsMu.Unlock()
+			continue
+		}
+		//metrics hit pool
+		return cn, nil
+	}
+
+	netconn, err := p.opt.Dialer(context.Background())
+	if err != nil {
+		p.sema.release()
+		return nil, err
+	}
+	p.connsMu.Lock()
+	defer p.connsMu.Unlock()
+
+	//Pool is full
+	if p.totalManagedConns > p.opt.PoolSize {
+		c := newUnmanagedConnFromNet(netconn) // will not be put back to pool later
+		return c, nil
+	} else {
+		c := newManagedConnFromNet(netconn)
+		p.totalManagedConns++
+		p.totalIdleConns++
+		p.conns[c.id] = c
+		return c, nil
+	}
+}
+
+func (p *ConnPool) Put(c *Conn) {
+	//TODO: check leftover read byte in socket
+	if !c.managed {
+		p.closeConn(c)
+		p.sema.release()
+		return
+	}
+	p.connsMu.Lock()
+	p.idleConns = append(p.idleConns, c)
+	p.totalIdleConns++
+	p.connsMu.Unlock()
+	p.sema.release()
+}
+
+func (p *ConnPool) closed() bool {
+	_, stillOpen := <-p.closedChan
+	return !stillOpen
+}
+
+func (p *ConnPool) connReaper() {
+	ticker := time.NewTicker(p.opt.IdleCheckFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if p.closed() {
+				return
+			}
+			p.reapStaleConns()
+		case <-p.closedChan:
+			return
+		}
+	}
+}
+
+//must used with lock
+func (p *ConnPool) ensureMinIdleConns() {
+	if p.opt.MinIdleConn < 0 {
+		return
+	}
+	totalErrCount := uint32(0)
+
+	for p.totalManagedConns < p.opt.PoolSize && p.totalIdleConns < p.opt.MinIdleConn {
+		p.totalManagedConns++ //TODO: is this counter necessary
+		p.totalIdleConns++
```

Comment:
should both these two counters increase atomically?

Reply:
ye, I think the counters related to conn should increase/decrease atomically since we don't know when a conn is taken or return

Reply:
in that case, should we use this?

https://golang.org/pkg/sync/atomic/

sorry, I'm not familiar with golang language, please advise

Reply:
I'm no golang expert either, but yeah using add functions from that package seems to be the way.

Reply:
All the code that call this function put it in a mutex lock/unlock block. I'll put some more comments for this

Reply:
is there any way to add an assert to this function, which would be:

- a no-op if we build with release mode
- check if the current goroutine is holding the mutex, if we build with debug mode

Reply:
That function now has 2 versions for release and debug. Debug version will check if the mutex is locked before it is called. If it is not, it will panic.
I have to use 2 different version of mutexes for release/debug as well.